Applications/spire/source/charting/cached_chart_model.cpp

```diff
@@ -36,90 +37,135 @@ QtPromise<std::vector<Candlestick>> CachedChartModel::load(ChartValue first,
       });
     }
   }
-  return load_data({first, last});
+  return load_data(get_gaps(first, last), first, last);
 }
 
 connection CachedChartModel::connect_candlestick_slot(
     const CandlestickSignal::slot_type& slot) const {
   return m_chart_model->connect_candlestick_slot(slot);
 }
 
+std::vector<ChartRange> CachedChartModel::get_gaps(ChartValue first,
+    ChartValue last) {
+  if(m_ranges.empty()) {
+    return {{ChartRange{first, last}}};
+  }
+  auto range_points = std::vector<ChartValue>({first, last});
+  for(auto& range : m_ranges) {
+    if(range.m_end < first) {
+      continue;
+    }
+    if(range.m_start >= last) {
+      break;
+    }
+    if(range.m_start <= first && range.m_end >= first) {
+      range_points.front() = range.m_end;
+    } else if(range.m_start <= last && range.m_end >= last) {
+      range_points.back() = range.m_start;
+    }
+  }
+  auto gaps = std::vector<ChartRange>();
+  for(auto i = std::size_t(0); i < range_points.size(); i += 2) {
+    gaps.push_back({range_points[i], range_points[i + 1]});
```

Comment:
I suspect there is a bug here.

Reply:
I forgot to add an else statement that was in the pseudocode that adds more points to the range_points vector.

Because m_loaded_data is erased between [A, B], where A and B represents the newly loaded data, this issue doesn't get caught in the tests; it just makes the cache less efficient. So the 'bug' here is that the cache is loading data it already has, then overwriting it when the new data is merged.

Reply:
Add a unit test and ensure your test fails, then implement the fix.

Reply:
I don't think I can write a test that fails for this because it will always work, it's just inefficient.